lib/SILGen/SILGenThunk.cpp

```diff
@@ -156,6 +156,13 @@ void SILGenFunction::emitCurryThunk(SILDeclRef thunk) {
   SILValue toFn = getNextUncurryLevelRef(*this, vd, thunk,
                                          selfArg, subs);
 
+  // A curry thunk is serialized only if the target callee is serialized.
+  if (auto *FRI = dyn_cast<FunctionRefInst>(toFn)) {
+    if (FRI->getReferencedFunction()->isSerialized() ==
```

Comment:
Cn you move this elsewhere into emitCurryThunk() etc?

Reply:
Sorry, I do not quite understand what you mean. It is in the emitCurryThunk already.

Reply:
What was the bug that prompted this change? I'm not convinced it is correct.

Remember that the serialized flag has three states:

- not serialized
- serializable
- serialized

"Serializable" means that it is serialized only if referenced from another serialized function. Since curry thunks have shared linkage, they need to be serialized if referenced from another serialized function (ie, the caller). It doesn't matter if the callee is serialized or not. Consider the case where the callee is public and not serialized; the thunk has shared linkage, which your change will make not serialized. Then if I use the curry thunk from a serialized caller, it will blow up.

```
public struct S {
  public func f() {}
}

@_inlineable public func g() {
  print(S.f)
}
```

Reply:
The bug was that the  curry thunk was marked serialized but the actual function that it calls was marked as NotSerialized.

Reply:
I'm all open to suggestions regarding how to fix it in a proper way.

Reply:
My naive idea was that if the target function is NotSerialized, then we cannot mark the curry thunk Serialized

Reply:
Can you come up with a reduced test case since this bug is not really related to your PR?

Serialized functions can reference non-serialized functions if they're public. So a curry thunk for anything public should be serializable since it can be serialized.

However, curry thunks for internal things should not be serializable. Was this the problem?

Reply:
May be. I need to reproduce it again to come up with a small test-case. I haven't created one when I hit it ;-(